### DISCUSSION TITLE
Explanation of container app target and image build?

### DISCUSSION THREAD

**Discussion by pamelafox** [Azure/azure-dev#1445] [Q&A] [answered]
We are working on various samples for Azure Container Apps, and we're struggling to understand how azd is able to deploy an image from the app code to ACR and reference that image for ACA in a single deploy. 

That seems to be what the TODO ACA apps do, as you can do a deploy and see the ACAs reference an image stored in the newly created ACR, but we don't understand what magic helps that happen. 

Relevant files:
https://github.com/Azure-Samples/todo-python-mongo-aca/blob/main/infra/main.parameters.json
https://github.com/Azure-Samples/todo-python-mongo-aca/blob/main/infra/main.bicep#L75
https://github.com/Azure-Samples/todo-python-mongo-aca/blob/main/infra/app/web.bicep#L35 
https://github.com/Azure-Samples/todo-python-mongo-aca/blob/main/infra/app/web.bicep#LL52

Do we need to set particular environment variable names? Do we need to structure the Bicep in a particular order, to make sure image exists first?

Thanks!

**Comment by weikanglim** [comment, marked as answer]
@pamelafox 


1. The name of the created ACR image is always saved into the environment as `SERVICE_<service name>_IMAGE_NAME`.
2. The current code deployment mechanism for ACA is via bicep, so create a bicep file for ACA deployment, and specify it in `azure.yaml`. [azure.yaml](https://github.com/Azure-Samples/todo-python-mongo-aca/blob/5381c011587f5ab904574c5eab978a835068f580/azure.yaml#L9), [bicep](https://github.com/Azure-Samples/todo-python-mongo-aca/blob/5381c011587f5ab904574c5eab978a835068f580/infra/app/web.bicep) examples. Note that this bicep file does not need to be referenced by `main.bicep`.
3. Create a `parameters.json` for your ACA deployment that uses `SERVICE_<service name>_IMAGE_NAME` as a parameter value substitution. [Example](https://github.com/Azure-Samples/todo-python-mongo-aca/blob/5381c011587f5ab904574c5eab978a835068f580/infra/app/web.parameters.json#L12).

**Reply by pamelafox** [reply]
Thank you! Giving it a go now.

**Reply by pamelafox** [reply]
Adding a few more notes based on things I missed:

* Ensure AZURE_CONTAINER_REGISTRY_ENDPOINT is set as an output of your infrastructure.
* Ensure you create a key vault and outputs its name, as a key vote is needed for the managed identity output (SERVICE_<service name)_IDENTITY_PRINCIPAL_ID)

Two examples I made:
* https://github.com/pamelafox/simple-flask-server-container (one image, no database)
* https://github.com/pamelafox/flask-surveys-container-app (one image, one database)

**Reply by weikanglim** [reply]
Good catch! `AZURE_CONTAINER_REGISTRY_ENDPOINT` is indeed required currently. I know the topic of list of well-known environment variables have come up often. I've started a discussion with my team with an initial draft of documentation here: #1458

I also know we're lacking proper documentation on `containerapp` authoring. Stay tuned as we might make tweaks around this area before publishing an official strategy.

**Comment by lambdakris** [comment]
Dang, could this be added to the documentation? I had similar questions about this seemingly magical behavior so that I could add new services to my project and modify the bicep and azure.yaml accordingly. I spent quite a bit of time scouring the docs, watching vids, and just tinkering...all to no avail, and finally stumbled across this issue only after I had given up all hope!

**Reply by pamelafox** [reply]
I've forwarded this thread to the azd docs team, they have some plans to document soon, I believe! Sorry you had to go through that, let me know if you still need help getting it set up. I have lots of examples here that do it: https://github.com/pamelafox

**Reply by lambdakris** [reply]
Wow, looking through your examples now, thanks so much!
